base/setup/usetup/usetup.c

```diff
@@ -626,6 +626,176 @@ UpdateKBLayout(VOID)
     }
 }
 
+/**
+ * @ProgressTimeOutStringHandler
+ * @implemented
+ *
+ * Handles the generation (displaying) of the timeout
+ * countdown to the screen dinamically
+ *
+ * @return
+ * FALSE on function termination
+ *
+ */
+static
+BOOLEAN NTAPI
+ProgressTimeOutStringHandler(
+    IN PPROGRESSBAR Bar,
+    IN BOOLEAN AlwaysUpdate,
+    OUT PSTR Buffer,
+    IN SIZE_T cchBufferSize)
+{
+    ULONG OldProgress = Bar->Progress;
+
+    if (Bar->StepCount == 0)
+    {
+        Bar->Progress = 0;
+    }
+    else
+    {
+        Bar->Progress = Bar->StepCount - Bar->CurrentStep;
+    }
+
+    if (Bar->ProgressFormatText &&
+        (AlwaysUpdate || (Bar->Progress != OldProgress)))
+    {
+        RtlStringCchPrintfA(Buffer, cchBufferSize,
+            Bar->ProgressFormatText, Bar->Progress);
+
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+/**
+ * @ProgressBarOnReboot
+ * @implemented
+ *
+ * Displays and draws a progress bar with red colour with a countdown.
+ * If the timeout has been reached, it'll go to on the flush page for reboot.
+ *
+ * @return
+ * Nothing
+ *
+ */
+static VOID
+ProgressBarOnReboot(
```

Comment:
This function seems quite complex for the simple task it is performing?

Reply:
Indeed, by looking at the function. Most of the function's code is used to manipulate the keyboard's input data. 
The previous function's code was very buggy and it did not handle the input's data properly (for example the progress bar was accelerating altogether with the countdown) each time you press a random key.

All the previous bugs are now gone with this code.